idea/src/org/jetbrains/kotlin/idea/inspections/ControlFlowWithEmptyBodyInspection.kt

```diff
@@ -21,11 +21,29 @@ class ControlFlowWithEmptyBodyInspection : AbstractKotlinInspection() {
         override fun visitIfExpression(expression: KtIfExpression) {
             val then = expression.then
             val elseKeyword = expression.elseKeyword
-            if (then.isEmptyBodyOrNull() && (elseKeyword == null || then?.hasComments() != true)) {
-                holder.registerProblem(expression, expression.ifKeyword)
-            }
-            if (elseKeyword != null && expression.`else`.isEmptyBodyOrNull()) {
-                holder.registerProblem(expression, elseKeyword)
+
+            val thenEmpty = then?.isEmptyBody()
+            val elseEmpty = expression.`else`?.isEmptyBody()
+
+            if (thenEmpty != false) {
+                // if () { }
+                if (elseEmpty != false) {
+                    // if () { } else { };
+                    holder.registerProblem(expression, expression.ifKeyword)
+                    if (elseKeyword != null) holder.registerProblem(expression, elseKeyword)
+                } else if (then?.hasComments() != true) {
+                    // if () { } else { foo() };
+                    holder.registerProblem(expression, expression.ifKeyword)
+                }
+            } else {
+                // if () { foo() }
+
+                if (elseKeyword == null) return
+
+                if (elseEmpty != false && expression.`else`?.hasComments() != true) {
+                    // if () { foo() } else {};
+                    holder.registerProblem(expression, elseKeyword)
```

Comment:
there are 2 branches - either problem in `then` and/or problem in `else` 

I think it would be more readable to have smth like
```
val thenProblem = ...
val elseProblem = ...

if (thenProblem) holder.registerProblem(expression, expression.ifKeyword)
if (elseProblem) holder.registerProblem(expression, elseKeyword)
```